main-extract.ts

```diff
@@ -507,63 +412,150 @@ export function extractFromTheseFiles(
+    thumb:     500 * thumbHeightFactor,                               // never above 300ms
+    filmstrip: 700 * numOfScreens * thumbHeightFactor,                // rarely above 15s, but 4K 30screens took 50s
+    clip:     1000 * clipSnippets * snippetLength * clipHeightFactor, // barely ever above 15s
+    clipThumb: 300 * clipHeightFactor,                                // never above 100ms
+  };
+}
+
+/**
+ * Return promise for whether file exists
+ * @param pathToFile string
+ */
+function checkFileExists(pathToFile: string): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    if (fs.existsSync(pathToFile)) {
+      return resolve(true);
+    } else {
+      return resolve(false);
+    }
+  });
+}
+
 /**
  * Replace original file with new file
  * use ffmpeg to convert and letterbox to fit width and height
  *
- * @param oldFile
- * @param newFile
+ * @param oldFile full path to thumbnail to replace
+ * @param newFile full path to sounce image to use as replacement
  * @param height
  */
 export function replaceThumbnailWithNewImage(
   oldFile: string,
   newFile: string,
   height: number
-) {
+): Promise<boolean> {
 
-  return new Promise((resolve, reject) => {
+  console.log('Resizing new image and replacing old thumbnail');
+
+  const width: number = Math.floor(height * (16 / 9));
+
+  const args = [
+    '-y', '-i', newFile,
+    '-vf', scaleAndPadString(width, height),
+    oldFile,
+  ];
+
+  return spawn_ffmpeg_and_run(args, 1000, 'replacing thumbnail');
+  // resizing an image file with ffmpeg should take less than 1 second
+}
+
+/**
+ * Generate the correct `scale=` & `pad=` string for ffmpeg
+ * @param width
+ * @param height
+ */
+function scaleAndPadString(width: number, height: number): string {
+  // sweet thanks to StackExchange!
+  // https://superuser.com/questions/547296/resizing-videos-with-ffmpeg-avconv-to-fit-into-static-sized-player
+
+  return 'scale=w=' + width + ':h=' + height + ':force_original_aspect_ratio=decrease,' +
+         'pad='     + width + ':'   + height + ':(ow-iw)/2:(oh-ih)/2';
 
-    console.log('Resizing new image and replacing old thumbnail');
+}
+
+/**
+ * Spawn ffmpeg and run the appropriate arguments
+ * Kill the process after maxRunningTime
+ * @param args            args to pass into ffmpeg
+ * @param maxRunningTime  maximum time to run ffmpeg
+ * @param description     log for console.log
+ */
+function spawn_ffmpeg_and_run(
+  args: string[],
+  maxRunningTime: number,
+  description: string
+): Promise<boolean> {
 
-    const width: number = Math.floor(height * (16 / 9));
+  return new Promise((resolve, reject) => {
 
-    const args = [
-      '-y', '-i', newFile,
-      '-vf', scaleAndPadString(width, height),
-      oldFile,
-    ];
+    // const t0: number = performance.now();
 
     const ffmpeg_process = spawn(ffmpegPath, args);
+
+    const killProcessTimeout = setTimeout(() => {
+      if (!ffmpeg_process.killed) {
```

Comment:
I'm unsure about this -- there's likely an official way to check if a process is still running 🤷‍♂